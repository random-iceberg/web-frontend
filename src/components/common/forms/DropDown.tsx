import React, { useState, useEffect, useRef } from "react";

/**
 * Props for the DropDown component.
 * @typedef {object} DropDownProps
 * @property {string} label - Text label displayed above the dropdown trigger.
 * @property {React.ReactNode} children - The options to display in the dropdown. Typically buttons or simple elements.
 * @property {string} id - Unique ID for the dropdown trigger button.
 * @property {string} value - The currently selected value (should match the content of one of the children options).
 * @property {(value: string) => void} onSelect - Callback function when an option is selected.
 * @property {boolean} [disabled=false] - Whether the dropdown is disabled.
 * @property {string} [description] - Optional description text displayed below the label.
 */
type DropDownProps = {
  label: string;
  children: React.ReactNode;
  id: string;
  value: string;
  onSelect: (value: string) => void;
  disabled?: boolean;
  description?: string; // Optional description text
};

/**
 * A custom dropdown component implemented with a button trigger and a panel showing options.
 * Handles opening/closing and selecting options.
 *
 * @component
 * @param {DropDownProps} props - Props for the DropDown component.
 * @example
 * const [selectedValue, setSelectedValue] = useState('');
 * const options = ['Option 1', 'Option 2', 'Option 3'];
 *
 * <DropDown
 *   id="my-dropdown"
 *   label="Select Option"
 *   value={selectedValue}
 *   onSelect={setSelectedValue}
 *   description="Choose one from the list."
 * >
 *   {options.map(opt => <button type="button" key={opt}>{opt}</button>)}
 * </DropDown>
 */
const DropDown: React.FC<DropDownProps> = ({
  label,
  children,
  id,
  value,
  onSelect,
  disabled = false,
  description,
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);
  const descriptionId = description ? `${id}-description` : undefined;

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        dropdownRef.current &&
        !dropdownRef.current.contains(event.target as Node)
      ) {
        setIsOpen(false);
      }
    };

    document.addEventListener("mousedown", handleClickOutside);
    return () => document.removeEventListener("mousedown", handleClickOutside);
  }, []);

  return (
    <div className="relative w-full" ref={dropdownRef}>
      <label
        htmlFor={id}
        className="block text-sm font-medium text-gray-700 mb-2"
      >
        {label}
      </label>
      <button
        type="button"
        id={id}
        onClick={() => !disabled && setIsOpen(!isOpen)}
        disabled={disabled}
        aria-haspopup="listbox"
        aria-expanded={isOpen}
        className="w-full text-left bg-white text-gray-700 text-base px-3 py-2 rounded-md border border-gray-300 shadow-sm focus:outline-none focus:ring-2 focus:ring-primary focus:border-primary disabled:bg-gray-100 disabled:text-gray-500 disabled:cursor-not-allowed"
        aria-describedby={descriptionId}
      >
        {value || label} {/* Display selected value or label as placeholder */}
      </button>
      {description && (
        <p id={descriptionId} className="mt-2 text-sm text-gray-600">
          {description}
        </p>
      )}
      {isOpen && (
        <div
          role="listbox"
          className="absolute mt-1 w-full bg-white border border-gray-300 rounded-md shadow-lg z-10 max-h-60 overflow-auto"
        >
          {React.Children.map(children, (child: any) => {
            // Assuming children are simple elements like <button> or <div> containing the value
            const childValue = child.props.children;
            return (
              <div
                role="option"
                aria-selected={value === childValue}
                onClick={() => {
                  onSelect(childValue);
                  setIsOpen(false);
                }}
                className={`px-4 py-2 cursor-pointer text-base text-gray-700 hover:bg-gray-100
                  ${value === childValue ? "bg-gray-200 font-semibold" : ""}`}
              >
                {/* Render the child directly (e.g., the button text) */}
                {child.props.children}
              </div>
            );
          })}
        </div>
      )}
    </div>
  );
};

export default DropDown; // Exporting as DropDown
