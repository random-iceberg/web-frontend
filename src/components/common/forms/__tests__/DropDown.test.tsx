import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import DropDown from "../DropDown";

describe("DropDown Component", () => {
  const mockOnSelect = jest.fn();
  const options = ["Option 1", "Option 2", "Option 3"];

  beforeEach(() => {
    mockOnSelect.mockClear();
  });

  it("renders correctly with a label and placeholder value", () => {
    const { asFragment } = render(
      <DropDown
        id="test-dropdown"
        label="Select an Option"
        value=""
        onSelect={mockOnSelect}
      >
        {options.map((opt) => (
          <button type="button" key={opt}>
            {opt}
          </button>
        ))}
      </DropDown>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with a selected value", () => {
    const { asFragment } = render(
      <DropDown
        id="test-dropdown"
        label="Select an Option"
        value="Option 2"
        onSelect={mockOnSelect}
      >
        {options.map((opt) => (
          <button type="button" key={opt}>
            {opt}
          </button>
        ))}
      </DropDown>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly when disabled", () => {
    const { asFragment } = render(
      <DropDown
        id="test-dropdown"
        label="Select an Option"
        value=""
        onSelect={mockOnSelect}
        disabled
      >
        {options.map((opt) => (
          <button type="button" key={opt}>
            {opt}
          </button>
        ))}
      </DropDown>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with a description", () => {
    const { asFragment } = render(
      <DropDown
        id="test-dropdown"
        label="Select an Option"
        value=""
        onSelect={mockOnSelect}
        description="This is a helpful description."
      >
        {options.map((opt) => (
          <button type="button" key={opt}>
            {opt}
          </button>
        ))}
      </DropDown>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("opens dropdown on click and shows options", () => {
    render(
      <DropDown
        id="test-dropdown"
        label="Select an Option"
        value=""
        onSelect={mockOnSelect}
      >
        {options.map((opt) => (
          <button type="button" key={opt}>
            {opt}
          </button>
        ))}
      </DropDown>,
    );
    // Initially, options are not visible
    expect(screen.queryByText("Option 1")).not.toBeInTheDocument();

    // Click the dropdown trigger (button with label text)
    fireEvent.click(screen.getByText("Select an Option"));

    // Now options should be visible
    expect(screen.getByText("Option 1")).toBeInTheDocument();
    expect(screen.getByText("Option 2")).toBeInTheDocument();
    expect(screen.getByText("Option 3")).toBeInTheDocument();
  });

  it("selects an option and calls onSelect when an option is clicked", () => {
    render(
      <DropDown
        id="test-dropdown"
        label="Select an Option"
        value=""
        onSelect={mockOnSelect}
      >
        {options.map((opt) => (
          <button type="button" key={opt}>
            {opt}
          </button>
        ))}
      </DropDown>,
    );

    // Open dropdown
    fireEvent.click(screen.getByText("Select an Option"));

    // Click on "Option 2"
    fireEvent.click(screen.getByText("Option 2"));

    expect(mockOnSelect).toHaveBeenCalledTimes(1);
    expect(mockOnSelect).toHaveBeenCalledWith("Option 2");
  });

  it("closes dropdown when an option is selected", () => {
    render(
      <DropDown
        id="test-dropdown"
        label="Select an Option"
        value=""
        onSelect={mockOnSelect}
      >
        {options.map((opt) => (
          <button type="button" key={opt}>
            {opt}
          </button>
        ))}
      </DropDown>,
    );
    fireEvent.click(screen.getByText("Select an Option")); // Open
    expect(screen.getByText("Option 1")).toBeInTheDocument();

    fireEvent.click(screen.getByText("Option 1")); // Select and close
    expect(screen.queryByText("Option 1")).not.toBeInTheDocument(); // Should be closed
  });

  it("closes dropdown when clicking outside", () => {
    render(
      <div>
        <DropDown
          id="test-dropdown"
          label="Select an Option"
          value=""
          onSelect={mockOnSelect}
        >
          {options.map((opt) => (
            <button type="button" key={opt}>
              {opt}
            </button>
          ))}
        </DropDown>
        <button type="button">Outside Button</button>
      </div>,
    );
    fireEvent.click(screen.getByText("Select an Option")); // Open
    expect(screen.getByText("Option 1")).toBeInTheDocument();

    // Click outside (on the body or another element)
    fireEvent.mouseDown(document.body); // Simulate mousedown on body

    expect(screen.queryByText("Option 1")).not.toBeInTheDocument(); // Should be closed
  });
});
