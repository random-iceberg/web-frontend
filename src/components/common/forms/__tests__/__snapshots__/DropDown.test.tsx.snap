// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown Component renders correctly when disabled 1`] = `
<DocumentFragment>
  <div
    class="relative w-full"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
      for="test-dropdown"
    >
      Select an Option
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="w-full text-left bg-white text-gray-700 text-base px-3 py-2 rounded-md border border-gray-300 shadow-sm focus:outline-none focus:ring-2 focus:ring-primary focus:border-primary disabled:bg-gray-100 disabled:text-gray-500 disabled:cursor-not-allowed"
      disabled=""
      id="test-dropdown"
      type="button"
    >
      Select an Option 
    </button>
  </div>
</DocumentFragment>
`;

exports[`DropDown Component renders correctly with a label and placeholder value 1`] = `
<DocumentFragment>
  <div
    class="relative w-full"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
      for="test-dropdown"
    >
      Select an Option
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="w-full text-left bg-white text-gray-700 text-base px-3 py-2 rounded-md border border-gray-300 shadow-sm focus:outline-none focus:ring-2 focus:ring-primary focus:border-primary disabled:bg-gray-100 disabled:text-gray-500 disabled:cursor-not-allowed"
      id="test-dropdown"
      type="button"
    >
      Select an Option 
    </button>
  </div>
</DocumentFragment>
`;

exports[`DropDown Component renders correctly with a selected value 1`] = `
<DocumentFragment>
  <div
    class="relative w-full"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
      for="test-dropdown"
    >
      Select an Option
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="w-full text-left bg-white text-gray-700 text-base px-3 py-2 rounded-md border border-gray-300 shadow-sm focus:outline-none focus:ring-2 focus:ring-primary focus:border-primary disabled:bg-gray-100 disabled:text-gray-500 disabled:cursor-not-allowed"
      id="test-dropdown"
      type="button"
    >
      Option 2 
    </button>
  </div>
</DocumentFragment>
`;

exports[`DropDown Component renders with a description 1`] = `
<DocumentFragment>
  <div
    class="relative w-full"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
      for="test-dropdown"
    >
      Select an Option
    </label>
    <button
      aria-describedby="test-dropdown-description"
      aria-expanded="false"
      aria-haspopup="listbox"
      class="w-full text-left bg-white text-gray-700 text-base px-3 py-2 rounded-md border border-gray-300 shadow-sm focus:outline-none focus:ring-2 focus:ring-primary focus:border-primary disabled:bg-gray-100 disabled:text-gray-500 disabled:cursor-not-allowed"
      id="test-dropdown"
      type="button"
    >
      Select an Option 
    </button>
    <p
      class="mt-2 text-sm text-gray-600"
      id="test-dropdown-description"
    >
      This is a helpful description.
    </p>
  </div>
</DocumentFragment>
`;
